File: sdk-android/src/main/java/cu/suitetecsa/sdk/android/balance/parser/BonusBalanceParser.java
package cu.suitetecsa.sdk.android.balance.parser;

import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import org.jetbrains.annotations.Contract;

import java.text.ParseException;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import cu.suitetecsa.sdk.android.balance.response.BonusBalance;
import cu.suitetecsa.sdk.android.model.BonusCredit;
import cu.suitetecsa.sdk.android.model.BonusData;
import cu.suitetecsa.sdk.android.model.BonusDataCU;
import cu.suitetecsa.sdk.android.model.BonusUnlimitedData;
import cu.suitetecsa.sdk.android.utils.StringUtils;

public class BonusBalanceParser {

    /**
     * Parses the bonus balance from a given CharSequence and returns a BonusBalance object.
     *
     * @return The parsed bonus balance as a BonusBalance object.
     */
    @RequiresApi(api = Build.VERSION_CODES.O)
    @NonNull
    @Contract("_ -> new")
    public static BonusBalance parseBonusBalance(CharSequence input) throws ParseException {
        // Regular expressions to match the bonus credit, bonus data, and bonus data CU patterns
        String creditPattern = "\\$(?<bonusCredit>([\\d.]+))\\s+vence\\s+(?<bonusCreditDueDate>(\\d{2}-\\d{2}-\\d{2}))\\.";
        Pattern creditRegex = Pattern.compile(creditPattern);
        Matcher creditMatcher = creditRegex.matcher(input);

        String dataPattern =
                "Datos:\\s+(ilimitados\\s+vence\\s+(?<unlimitedData>(\\d{2}-\\d{2}-\\d{2}))\\.)?" +
                        "(\\s+)?((?<dataAllNetworkBonus>(\\d+(\\.\\d+)?)(\\s)*([GMK])?B))?\\s+\\+\\s+" +
                        "((?<bonusDataLte>(\\d+(\\.\\d+)?)(\\s)*([GMK])?B)\\s+LTE)?\\s+vence\\s+" +
                        "((?<bonusDataDueDate>(\\d{2}-\\d{2}-\\d{2})))\\.";
        Pattern dataRegex = Pattern.compile(dataPattern);
        Matcher dataMatcher = dataRegex.matcher(input);

        String dataCUPattern =
                "Datos\\.cu\\s+(?<bonusDataCu>(\\d+(\\.\\d+)?)(\\s)*([GMK])?B)?\\s+vence\\s+" +
                        "(?<bonusDataCuDueDate>(\\d{2}-\\d{2}-\\d{2}))\\.";
        Pattern dataCURegex = Pattern.compile(dataCUPattern);
        Matcher dataCUMatcher = dataCURegex.matcher(input);

        BonusCredit bonusCredit = null;
        BonusData bonusData = null;
        BonusDataCU bonusDataCU = null;
        BonusUnlimitedData bonusUnlimitedData = null;

        if (creditMatcher.find()) {
            // Parse bonus credit
            bonusCredit = new BonusCredit(
                    Float.parseFloat(Objects.requireNonNull(creditMatcher.group("bonusCredit"))),
                    StringUtils.toDateMillis(creditMatcher.group("bonusCreditDueDate"))
            );
        }

        if (dataMatcher.find()) {
            // Parse bonus data and bonus unlimited data
            String dataAllNetworkBonus = dataMatcher.group("dataAllNetworkBonus");
            String bonusDataLte = dataMatcher.group("bonusDataLte");
            String bonusDataDueDate = dataMatcher.group("bonusDataDueDate");
            bonusData = new BonusData(
                    dataAllNetworkBonus != null ? StringUtils.toBytes(dataAllNetworkBonus) : null,
                    bonusDataLte != null ? StringUtils.toBytes(bonusDataLte) : null,
                    StringUtils.toDateMillis(bonusDataDueDate)
            );

            String unlimitedData = dataMatcher.group("unlimitedData");
            if (unlimitedData != null) {
                bonusUnlimitedData = new BonusUnlimitedData(StringUtils.toDateMillis(unlimitedData));
            }
        }

        if (dataCUMatcher.find()) {
            // Parse bonus data CU
            bonusDataCU = new BonusDataCU(
                    StringUtils.toBytes(Objects.requireNonNull(dataCUMatcher.group("bonusDataCu"))),
                    StringUtils.toDateMillis(dataCUMatcher.group("bonusDataCuDueDate"))
            );
        }

        // Return the parsed bonus balance
        return new BonusBalance(bonusCredit, bonusUnlimitedData, bonusData, bonusDataCU);
    }
}

File: sdk-android/src/main/java/cu/suitetecsa/sdk/android/balance/parser/BonusBalanceParser.kt
package cu.suitetecsa.sdk.android.balance.parser

import cu.suitetecsa.sdk.android.balance.response.BonusBalance
import cu.suitetecsa.sdk.android.model.BonusCredit
import cu.suitetecsa.sdk.android.model.BonusData
import cu.suitetecsa.sdk.android.model.BonusUnlimitedData
import cu.suitetecsa.sdk.android.model.DataCu
import cu.suitetecsa.sdk.android.model.Sms
import cu.suitetecsa.sdk.android.model.Voice
import cu.suitetecsa.sdk.android.utils.LongUtils.toRemainingDays
import cu.suitetecsa.sdk.android.utils.StringUtils
import cu.suitetecsa.sdk.android.utils.StringUtils.toBytes
import cu.suitetecsa.sdk.android.utils.StringUtils.toDateMillis

object BonusBalanceParser {
    /**
     * Parses the input text to extract bonus balance information, including credit, data, and specific data bonuses.
     *
     * @param input CharSequence containing the text to parse.
     * @return BonusBalance object containing parsed bonus credit, data, and other specific bonuses.
     */
    @JvmStatic
    fun extractBonusBalance(input: CharSequence): BonusBalance {
        return BonusBalance(
            input.extractCredit(),
            input.extractUnlimitedData(),
            input.extractData(),
            input.extractDataCu(),
            input.extractVoice(),
            input.extractSms()
        )
    }
}

val CharSequence.asBonusBalance: BonusBalance?
    get() = BonusBalance(
        this.extractCredit(),
        this.extractUnlimitedData(),
        this.extractData(),
        this.extractDataCu(),
        this.extractVoice(),
        this.extractSms()
    )

private fun CharSequence.extractCredit() =
    """\$(?<volume>([\d.]+))\s+vence\s+(?<dueDate>(\d{2}-\d{2}-\d{2}))\.""".toRegex().find(this)?.let {
        BonusCredit(
            it.groups["volume"]!!.value.toDouble(),
            toRemainingDays(toDateMillis(it.groups["dueDate"]!!.value))
        )
    }

private fun CharSequence.extractData(): BonusData? {
    val dataCountPattern = """(\d+(\.\d+)?)(\s)*([GMK])?B"""
    val dataCountGroup = """(?<volume>${dataCountPattern})"""
    val dataCountLtePattern = """(?<volumeLte>${dataCountPattern})\s+LTE"""
    val dataDueDatePattern = """(?<dueDate>(\d{2}-\d{2}-\d{2}))"""
    val fullDataCountPattern =
        """(${dataCountGroup})?(\s+)?(\+)?(\s+)?(${dataCountLtePattern})?\s+vence\s+${dataDueDatePattern}(\.)?"""
    val unlimitedDataPattern = """ilimitados\s+vence\s+(?<unlimitedData>(\d{2}-\d{2}-\d{2}))\."""
    val dataRegex = """Datos:\s+($unlimitedDataPattern)?(\s+)?($fullDataCountPattern)?""".toRegex()

    return dataRegex.find(this)?.let {
        val volume = it.groups["volume"]?.value
        val volumeLte = it.groups["volumeLte"]?.value
        val dueDate = it.groups["dueDate"]!!.value
        BonusData(
            volume?.let { dv -> toBytes(dv) },
            volumeLte?.let { dlv -> toBytes(dlv) },
            toRemainingDays(toDateMillis(dueDate))
        )
    }
}

private fun CharSequence.extractUnlimitedData() =
    """ilimitados\s+vence\s+(?<dueDate>(\d{2}-\d{2}-\d{2}))\."""
        .toRegex().find(this)?.let {
            it.groups["dueDate"]?.value?.let { ud ->
                BonusUnlimitedData(toRemainingDays(toDateMillis(ud)))
            }
        }

private fun CharSequence.extractDataCu() =
    """Datos\.cu\s+(?<volume>(\d+(\.\d+)?)(\s)*([GMK])?B)?\s+vence\s+(?<dueDate>(\d{2}-\d{2}-\d{2}))\."""
        .toRegex().find(this)?.let {
            DataCu(
                toBytes(it.groups["volume"]!!.value),
                toRemainingDays(toDateMillis(it.groups["dueDate"]!!.value))
            )
        }

private fun CharSequence.extractVoice() =
    """Voz:\s+(?<volume>(\d{2}:\d{2}:\d{2}))\s+vence\s+(?<dueDate>(\d{2}-\d{2}-\d{2}))\."""
        .toRegex().find(this)?.let {
            Voice(
                StringUtils.toSeconds(it.groups["volume"]!!.value),
                toRemainingDays(toDateMillis(it.groups["dueDate"]!!.value))
            )
        }

private fun CharSequence.extractSms() =
    """SMS:\s+(?<volume>(\d+))\s+vence\s+(?<dueDate>(\d{2}-\d{2}-\d{2}))\."""
        .toRegex().find(this)?.let {
            Sms(it.groups["volume"]!!.value.toLong(), toRemainingDays(toDateMillis(it.groups["dueDate"]!!.value)))
        }
